### PR TITLE
Contracts & Harnesses for widening_mul

### PR DESCRIPTION
**Pull request by rajathkotyal**
Towards : issue #59

Parent branch : main 

Revalidation : 
Per the discussion in https://github.com/model-checking/verify-rust-std/issues/59, we have to build and run Kani from feature/verify-rust-std branch.
To revalidate the verification results, run the following command. <harness_to_run> can be either num::verify to run all harnesses or num::verify::<harness_name> (e.g. widening_mul_u16_small) to run a specific harness.
```
kani verify-std  "path/to/library" \
    --harness <harness_to_run> \
    -Z unstable-options \
    -Z function-contracts \
    -Z mem-predicates
```

### CODE REVIEW

File: library/core/src/num/mod.rs
Comment (rajathkotyal) [reply library/core/src/num/mod.rs]: Thanks for pointing that @celinval , upon checking the function, we already have the overflowing precondition for the unchecked_mul function : 
```
#[requires(!self.overflowing_mul(rhs).1)]
pub const unsafe fn unchecked_mul(self, rhs: Self) -> Self 
```

And the widening_mul calls the unchecked_mul, Hence we can safely remove this condition as its already being checked at function level. Updated code.

Comment (celinval) [reply library/core/src/num/mod.rs]: Sorry, I forgot to mark the other assumptions above. Ideally, the harness shouldn't have any assumptions. Unless you bump into some issue like performance.

Comment (rajathkotyal) [reply library/core/src/num/mod.rs]: Yes @celinval , The assumptions are added because of the performance constraint in multiplication for u16, u32 and u64 types. u8 shouldn't have an issue since the total number of checks is around 256 × 256 = 65,536.

But for u16, its around 65,536 × 65,536 ~ 4.29 billion, which is infeasible, hence have added the assumptions.

Comment (celinval) [reply library/core/src/num/mod.rs]: Sorry... I totally missed that!

**Review by celinval — APPROVED** [review]

**Comment by rajathkotyal** [comment]
Thanks @celinval! , Need another approval on this before merge. @carolynzech could you please review it?

**Review by carolynzech — APPROVED** [review]

**Pull request merged**
